crucible-llvm/src/Lang/Crucible/LLVM/SimpleLoopFixpoint.hs

```diff
@@ -0,0 +1,748 @@
+  , fixpointAssumptionFrameIdentifier :: C.FrameIdentifier
+  , fixpointSubstitution :: MapF (W4.SymExpr sym) (FixpointEntry sym)
+  , fixpointRegMap :: C.RegMap sym args
+  , fixpointMemSubstitution :: Map (Natural, Natural, Natural) (MemFixpointEntry sym, C.StorageType)
+  , fixpointLoopCondition :: W4.Pred sym
+  , fixpointLoopIndexBound :: LoopIndexBound sym
+  }
+
+type FixpointMonad sym = StateT (MapF (W4.SymExpr sym) (FixpointEntry sym)) IO
+
+
+joinRegEntries ::
+  (?logMessage :: String -> IO (), C.IsSymInterface sym) =>
+  sym ->
+  Ctx.Assignment (C.RegEntry sym) ctx ->
+  Ctx.Assignment (C.RegEntry sym) ctx ->
+  FixpointMonad sym (Ctx.Assignment (C.RegEntry sym) ctx)
+joinRegEntries sym = Ctx.zipWithM (joinRegEntry sym)
+
+joinRegEntry ::
+  (?logMessage :: String -> IO (), C.IsSymInterface sym) =>
+  sym ->
+  C.RegEntry sym tp ->
+  C.RegEntry sym tp ->
+  FixpointMonad sym (C.RegEntry sym tp)
+joinRegEntry sym left right = case C.regType left of
+  C.LLVMPointerRepr w
+    | List.isPrefixOf "cmacaw_reg" (show $ W4.printSymNat $ C.llvmPointerBlock (C.regValue left))
+    , List.isPrefixOf "cmacaw_reg" (show $ W4.printSymExpr $ C.llvmPointerOffset (C.regValue left)) -> do
+      liftIO $ ?logMessage "SimpleLoopFixpoint.joinRegEntry: cmacaw_reg"
+      return left
+    | C.llvmPointerBlock (C.regValue left) == C.llvmPointerBlock (C.regValue right) -> do
+      liftIO $ ?logMessage "SimpleLoopFixpoint.joinRegEntry: LLVMPointerRepr"
+      subst <- get
+      if isJust (W4.testEquality (C.llvmPointerOffset (C.regValue left)) (C.llvmPointerOffset (C.regValue right)))
+      then do
+        liftIO $ ?logMessage "SimpleLoopFixpoint.joinRegEntry: LLVMPointerRepr: left == right"
+        return left
+      else case MapF.lookup (C.llvmPointerOffset (C.regValue left)) subst of
+        Just join_entry -> do
+          liftIO $ ?logMessage $
+            "SimpleLoopFixpoint.joinRegEntry: LLVMPointerRepr: Just: "
+            ++ show (W4.printSymExpr $ bodyValue join_entry)
+            ++ " -> "
+            ++ show (W4.printSymExpr $ C.llvmPointerOffset (C.regValue right))
+          put $ MapF.insert
+            (C.llvmPointerOffset (C.regValue left))
+            (join_entry { bodyValue = C.llvmPointerOffset (C.regValue right) })
+            subst
+          return left
+        Nothing -> do
+          liftIO $ ?logMessage "SimpleLoopFixpoint.joinRegEntry: LLVMPointerRepr: Nothing"
+          join_varaible <- liftIO $ W4.freshConstant sym (userSymbol' "reg_join_var") (W4.BaseBVRepr w)
+          let join_entry = FixpointEntry
+                { headerValue = C.llvmPointerOffset (C.regValue left)
+                , bodyValue = C.llvmPointerOffset (C.regValue right)
+                }
+          put $ MapF.insert join_varaible join_entry subst
+          return $ C.RegEntry (C.LLVMPointerRepr w) $ C.LLVMPointer (C.llvmPointerBlock (C.regValue left)) join_varaible
+    | otherwise ->
+      fail $
```

Comment:
Is this a `panic` situation?

Reply:
I used `panic` for situations that shouldn't happen, and `fail` for situations that can happen but are not supported. It's possible to handle pointers with symbolic base, it's just out of scope.

Reply:
I have been trying to migrate to throwing an `Unsupported` sim error in these situations.

Reply:
I recently added a `throwUnsupported` combinator that makes it easier to do this.